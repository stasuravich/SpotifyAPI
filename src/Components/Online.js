import React, {useState, useEffect, memo, useRef} from 'react';
import '../Css/Online.css';
import axios from 'axios';

const Online = memo(props=> {
  console.log("Online Component");
  const [open, setOpen]=useState(false);
  const [onlineClicked, setOnlineClicked]=useState(false);
  const addingTrack=useRef();
  const userInfo=useRef();
  let curTracks;

  useEffect(()=> {
    axios('https://api.spotify.com/v1/me', {
        method: 'GET',
        headers: {'Authorization' : 'Bearer ' + props.logInfo}
      })
      .then((response) => {
        userInfo.current=response.data;
      })
  },[props.logInfo]);

  useEffect(() => {
    const handleClickInside = event=>{
      if(open && props.inputRef.current.contains(event.target)){
        setOpen(false);
      }
      else if(props.inputRef.current.contains(event.target)){
        setOnlineClicked(true);
      }
      else if (document.activeElement.className!=="AddButton"){
        setOnlineClicked(false);
        if(open){
          setOpen(false);
        }
      }
    }
    document.addEventListener("click", handleClickInside);
    return () => {
      document.removeEventListener("click", handleClickInside);
    };
  }, [open])

  const addSongClicked=async addingId=>{
    let addSong=false;
    curTracks = await props.getPlaylist(0, curTracks, addingId);
    for(const item of curTracks){
      if(item.track.uri===addingTrack.current){
        alert("Song already in the playlist")
        break;
      }
      else if(item===curTracks[curTracks.length-1]){
        addSong=true;
      }
    }
    if(addSong){
      curTracks=null;
      axios(`https://api.spotify.com/v1/playlists/${addingId}/tracks`, {
        method: "POST",
        headers: {'Authorization' : 'Bearer ' + props.logInfo},
        data: { "uris": [addingTrack.current]}
      })
      .then (async _ => {
        curTracks = await props.getPlaylist(0, curTracks, addingId);
        if(addingId===props.playlists.current.selectedPlaylist){
          props.setDispPlaylist(curTracks);
          if(props.playlistTrack.current===props.playlists.current.selectedPlaylist){
            props.setTracks({selectedTrack: props.tracks.selectedTrack, listOfTracksFromApi: curTracks});
          }
        }
        else if(props.playlistTrack.current===addingId){
          props.setTracks({selectedTrack: props.tracks.selectedTrack, listOfTracksFromApi: curTracks});
        }
      })
    }
  }

  const seePlaylists=e=>{
    document.getElementById("PopUp").style.marginTop=e.currentTarget.id+"px";
    if(!open || (addingTrack.current && addingTrack.current===e.currentTarget.value)){
      setOpen(!open);
    }
    addingTrack.current=e.currentTarget.value;
  }

  const addTheSong=e=> {
    addSongClicked(e.target.id);
  };

  return (onlineClicked && props.songs ?(
    <div className="AboveOnline">
      <div className= "Online">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"/ >
        {typeof props.songs ==="object" ?
          props.songs.tracks.items.map((item, idx) =>
            <div className="AddParent" key={idx}><button id={idx*23+57} value = {item.uri} onClick={seePlaylists} className="AddButton">
              <i className="fa fa-plus"></i><div className="AddMessage" >&nbsp;Add to playlist</div></button>{item.name}: {item.artists[0].name}</div>) :
          <div> '{props.songs}' does not exist </div>}
      </div>
      <div id ="PopUp" className="PopUp">
      {open && props.playlists.current.listOfPlaylistsFromAPI.map((item, idx) => userInfo.current.display_name===item.owner.display_name &&
        <button className="AddToPlay" key={idx} id = {item.id} onClick={addTheSong}>{item.name}</button>
      )}
      </div>
    </div>): null
  );
})
export default Online;
