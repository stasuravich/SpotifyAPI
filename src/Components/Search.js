import React, {useRef, memo} from 'react';
import '../Css/Search.css';

const Search = memo(function Search(props){
  //console.log(props.class);
  const inputRef = useRef();

  props.alerter(inputRef);

  return(
    <div className= {props.class} >
      <input ref = {inputRef} type="text" placeholder = {props.placeh} onChange={(e) => props.setQuery(e.target.value)}></input>
    </div>
  )
})

export default Search;
