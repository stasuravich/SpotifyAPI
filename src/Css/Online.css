.AboveOnline{
  float: right;
  width: 420px;

}
.Online{
  position: fixed;
  margin-top: 29px;
  top: 0;
  background: #FFF5EE;
  width: 420px;
  z-index: 20;
}

.PopUp{
  position: fixed;
  margin-left: 85px;
  margin-top: 35px;
  top: 0;
  z-index: 30;
  width: 230px;
  display: flex;
  flex-direction: column;
}

.AddToPlay{
  text-align: left;
}

.AddButton{
  position: absolute;
  margin-right: 1px;
  top: 2;
  right: 0;
  border-width: thin;
}

.AddMesage{
  margin: auto;
}
.AddParent .AddButton{
  display: none;
}
.AddParent:hover .AddButton{
  display: inline-block;
}
.AddButton .AddMesage{
  display: none;
}
.AddButton:hover .AddMesage{
  display: inline-block;
}
