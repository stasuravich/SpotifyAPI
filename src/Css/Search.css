.SearchPlaylist{
  padding-top: 10px;
  order: 1;
  padding-left: 250px;
  visibility: hidden;
}

.SearchOnline{
  padding-top: 10px;
  order: 3;
  padding-right: 250px;
}
