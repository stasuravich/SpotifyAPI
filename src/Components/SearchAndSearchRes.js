import React, {useState, useRef, memo, useCallback} from 'react';
import Online from './Online';
import Search from './Search';

import axios from 'axios';

const SearchAndSearchRes = memo(props=>{
  const [searchRes, setSearchRes] = useState();
  const inputRef = useRef();

  const changeSearchRes = useCallback(query =>{
    if(query){
      axios(`https://api.spotify.com/v1/search?limit=15&type=track&q=${query}`, {
          method: 'GET',
          headers: {'Authorization' : 'Bearer ' + props.logInfo}
        })
        .then((response) => {
          if(response.data.tracks.items.length){
            setSearchRes(response.data);
          }
          else{
            setSearchRes(query);
          }
        })
    }
    else{
      setSearchRes(null);
    }
  }, [searchRes])

  return(
    <>
      <Search class="SearchOnline" placeh="Search for a song..." inputRef={inputRef} changeSearchRes={changeSearchRes}/>
      <Online songs={searchRes} logInfo={props.logInfo} playlists = {props.playlists} getPlaylist={props.getPlaylist} playlistTrack={props.playlistTrack} setTracks={props.setTracks} tracks={props.tracks} setDispPlaylist={props.setDispPlaylist} inputRef={inputRef} />
    </>
  )
})

export default SearchAndSearchRes;
