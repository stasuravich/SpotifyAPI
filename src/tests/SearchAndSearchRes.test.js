import React from 'react';
import SearchAndSearchRes from '../Components/SearchAndSearchRes';
import Search from '../Components/Search';
import Online from '../Components/Online';
import {shallow, mount} from "enzyme";
import toJson from "enzyme-to-json";

const searchClass="SearchOnline";
const tracks = {selectedTrack: "Latency", listOfTracksFromApi: ["Latency", "Remember", "Error 404"]};

describe("rendering components", ()=> {
  it("renders SearchRes without crashing", ()=>{
    shallow(<SearchAndSearchRes/>);
  });
})

describe("passing props", ()=>{
  const onlineWrapper=mount(<Online tracks={tracks} />);
  const searchWrapper=mount(<Search class={searchClass}/>);
  it("accepts Online props", () => {
    expect(onlineWrapper.props().tracks).toEqual(tracks);
  })
  it("accepts Search prop", ()=>{
    expect(searchWrapper.props().class).toEqual(searchClass);
  })
})

describe("snapshots", () => {
  it("SearchRes snapshots", () => {
    const tree= shallow(<SearchAndSearchRes/>);
    expect(toJson(tree)).toMatchSnapshot();
  });
})
