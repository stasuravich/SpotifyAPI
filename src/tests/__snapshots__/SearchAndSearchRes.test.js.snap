// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots SearchRes snapshots 1`] = `
<Fragment>
  <Memo()
    changeSearchRes={[Function]}
    class="SearchOnline"
    inputRef={
      Object {
        "current": undefined,
      }
    }
    placeh="Search for a song..."
  />
  <Memo()
    inputRef={
      Object {
        "current": undefined,
      }
    }
  />
</Fragment>
`;
